src/edge/simulator.ts

```diff
@@ -91,85 +56,85 @@ export class Simulator {
+        return (Simulator.currentPlatform === "linux" || Simulator.currentPlatform === "darwin") ? `sudo ${command}` : command;
+    }
 
-            let input: vscode.MessageItem | undefined;
-            if (force) {
-                input = await vscode.window.showWarningMessage(message, { modal: true }, ...items);
-            } else {
-                input = await vscode.window.showWarningMessage(message, ...items);
-            }
+    private isInstalling: boolean = false;
+    private latestStandaloneSimulatorInfoJson: any;
 
-            if (input === install) {
-                try {
-                    await Executor.executeCMD(outputChannel, "pip", {shell: true}, "install --upgrade iotedgehubdev");
-                } catch (error) {
-                    if (outputChannel) {
-                        outputChannel.appendLine(`${Constants.failedInstallSimulator} ${error.message}`);
-                    }
-                    ret =  InstallReturn.Failed;
-                }
-            } else {
-                ret = InstallReturn.Canceled;
-            }
-        } else {
-            ret = InstallReturn.NoPip;
-        }
-        return ret;
+    constructor(private context: vscode.ExtensionContext) {
     }
 
-    private static async simulatorInstalled(): Promise<boolean> {
-        return await Simulator.checkCmdExist("iotedgehubdev");
-    }
+    public async validateSimulatorUpdated(outputChannel: vscode.OutputChannel = null): Promise<void> {
+        let message: string;
+        let type: string = "";
+        const telemetryName = "simulatorUpdated";
 
-    private static async checkPipInstalled(): Promise<boolean> {
-        return await Simulator.checkCmdExist("pip");
-    }
+        const simulatorType = await this.getSimulatorType();
+        if (simulatorType === SimulatorType.NotInstalled) {
+            message =  Constants.needSimulatorInstalledMsg;
+            type = "install";
+        } else {
+            if (Simulator.currentPlatform === "win32" && this.getUserConfiguredVersion()) {
+                return;
+            }  else {
+                const version: string | null = await this.getCurrentSimulatorVersion();
+                if (version && semver.valid(version)) {
+                    const latestVersion: string | undefined = await this.getLatestSimulatorVersion();
+                    if (latestVersion && semver.gt(latestVersion, version)) {
+                        message = `${Constants.updateSimulatorMsg} (${version} to ${latestVersion})`;
+                    }
+                } else {
+                    message = Constants.updateSimulatorMsg;
+                }
 
-    private static async checkCmdExist(cmd: string): Promise<boolean> {
-        try {
-            await Executor.executeCMD(null, cmd, { shell: true }, "--version");
-            return true;
-        } catch (error) {
-            return false;
+                if (simulatorType === SimulatorType.Pip) {
```

Comment:
so the logic is inconsistent here. If no message type should also not assigned.

Reply:
OK, I think we can use return to avoid unnecessary code